src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs

```diff
@@ -408,6 +408,62 @@ internal void ThrowIfCompletedUnsuccessfully()
+        /// <returns>The <see cref="ValueTask"/> representing the asynchronous wait.  It may or may not be the same instance as the current instance.</returns>
+        /// <remarks>
+        /// Canceling the <see cref="WaitAsync"/> doesn't affect the execution of the original task itself; if a <see cref="WaitAsync"/> is canceled on a task that
+        /// hasn't yet completed, the original task may continue to execute even after the <see cref="WaitAsync"/> task has completed.
+        /// Once <see cref="WaitAsync"/> has been called, the original <see cref="ValueTask"/> shouldn't be accessed again, regardless of
+        /// whether the operation completes or is canceled.
+        /// </remarks>
+        public ValueTask WaitAsync(CancellationToken cancellationToken) =>
+            WaitAsync(Timeout.InfiniteTimeSpan, cancellationToken);
+
+        /// <summary>Gets a <see cref="ValueTask"/> that will complete when this <see cref="ValueTask"/> completes or when the specified timeout expires.</summary>
+        /// <param name="timeout">The timeout after which the <see cref="ValueTask"/> should be faulted with a <see cref="TimeoutException"/> if it hasn't otherwise completed.</param>
+        /// <returns>The <see cref="Task"/> representing the asynchronous wait.  It may or may not be the same instance as the current instance.</returns>
+        /// <remarks>
+        /// Timeouts don't <see cref="WaitAsync"/> affect the execution the original task itself; if a <see cref="WaitAsync"/> is canceled or times out on a task that
+        /// hasn't yet completed, the original task may continue to execute even after the <see cref="WaitAsync"/> task has completed.
+        /// Once <see cref="WaitAsync"/> has been called, the original <see cref="ValueTask"/> shouldn't be accessed again, regardless of
+        /// whether the operation completes or times out.
+        /// </remarks>
+        public ValueTask WaitAsync(TimeSpan timeout) =>
+            WaitAsync(timeout, default);
+
+        /// <summary>Gets a <see cref="ValueTask"/> that will complete when this <see cref="ValueTask"/> completes, when the specified timeout expires, or when the specified <see cref="CancellationToken"/> has cancellation requested.</summary>
+        /// <param name="timeout">The timeout after which the <see cref="ValueTask"/> should be faulted with a <see cref="TimeoutException"/> if it hasn't otherwise completed.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for a cancellation request.</param>
+        /// <returns>The <see cref="Task"/> representing the asynchronous wait.  It may or may not be the same instance as the current instance.</returns>
+        /// <remarks>
+        /// Canceling or timing out the <see cref="WaitAsync"/> doesn't affect the execution of the original task itself; if a <see cref="WaitAsync"/> is canceled or times out on a task that
```

Comment:
Might be worth adding a note stating that any call to `WaitAsync()` means that the source task cannot be awaited (even if the returned `ValueTask` is never awaited itself). As such it might be impossible to obtain the result of the source task if the waiter task times out.

Likewise, `WaitAsync()` must never be called twice on the same `ValueTask` instance.

Reply:
That's what I was trying to say with the:
```
        /// Once <see cref="WaitAsync"/> has been called, the original <see cref="ValueTask"/> shouldn't be accessed again, regardless of
        /// whether the operation completes, times out, or is canceled.
```
You're suggesting I just further spell out the implications of that?

Reply:
On second thought, that wording probably covers it.

Reply:
Ok